### PR TITLE
Fix flax gpt2 hidden states

### LINKED ISSUE #13102: inconsistency of the last element in hidden_states between PyTorch/Flax GPT2(Neo)
### Who can help
@patrickvonplaten  @patil-suraj 

## Information

The current Flax version of GPT2/GPTNeo give different results for the last element in `hidden_states` if `output_hidden_states=True`. This difference comes from the following fact:

In Flax GPT2 (and GPTNeo similarly), `all_hidden_states` is prepared in `FlaxGPT2BlockCollection` which has no layer norm layer (`ln_f`), therefore the last hidden state is added before applying layer normalization.
While in PyTorch/TF GPT2, it is prepared in `GPT2Model` or `TFGPT2MainLayer`, which contain `ln_f` layer, and the last hidden state is added after applying layer normalization.

This could be fixed by updating the outputs in `FlaxGPT2Module.__call__`, (if it's worth the change), something like

```
        hidden_states = outputs[0]
        hidden_states = self.ln_f(hidden_states)
        
        all_hidden_states = None
        if output_hidden_states:
            if not return_dict:
                all_hidden_states = outputs[1]
            else:
                all_hidden_states = outputs.hidden_states
            all_hidden_states = all_hidden_states[:-1] + (hidden_states,)

        if not return_dict:
            if all_hidden_states:
                return (hidden_states, all_hidden_states) + outputs[2:]
            else:
                return (hidden_states,) + outputs[1:]

        return FlaxBaseModelOutputWithPastAndCrossAttentions(
            last_hidden_state=hidden_states,
            hidden_states=all_hidden_states,
            attentions=outputs.attentions,
            cross_attentions=outputs.cross_attentions,
        )
``` 

### Related places in the source code

PyTroch GPT2
https://github.com/huggingface/transformers/blob/773d386041b2761204dcc67b316904d8d5b412da/src/transformers/models/gpt2/modeling_gpt2.py#L820

```
        hidden_states = self.ln_f(hidden_states)
        ...
        # Add last hidden state
        if output_hidden_states:
            all_hidden_states = all_hidden_states + (hidden_states,)
```

TensorFlow GPT2
https://github.com/huggingface/transformers/blob/773d386041b2761204dcc67b316904d8d5b412da/src/transformers/models/gpt2/modeling_tf_gpt2.py#L397

```
        hidden_states = self.ln_f(hidden_states)
        ...
        # Add last hidden state
        if inputs["output_hidden_states"]:
            all_hidden_states = all_hidden_states + (hidden_states,)
```

Flax GPT2
https://github.com/huggingface/transformers/blob/773d386041b2761204dcc67b316904d8d5b412da/src/transformers/models/gpt2/modeling_flax_gpt2.py#L461

```
       # In `FlaxGPT2BlockCollection` which has no `ln_f` (only exist in `FlaxGPT2Module`)
     
        if output_hidden_states:
            all_hidden_states += (hidden_states,)

```

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

Fixes #13102 

## Before submitting

- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests), Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? 
    [ inconsistency of the last element in hidden_states between PyTorch/Flax GPT2(Neo) #13102 ](https://github.com/huggingface/transformers/issues/13102#issuecomment-897687182)

## Who can review?

@patrickvonplaten @patil-suraj

### CODE REVIEW

File: src/transformers/models/gpt2/modeling_flax_gpt2.py [hunk 1]
```diff
@@ -527,12 +524,23 @@ def __call__(
...
         hidden_states = self.ln_f(hidden_states)
 
+        all_hidden_states = None
+        if output_hidden_states:
+            if not return_dict:
+                all_hidden_states = outputs[1]
+            else:
+                all_hidden_states = outputs.hidden_states
+            all_hidden_states = all_hidden_states + (hidden_states,)
+
```
Comment (patil-suraj) [inline src/transformers/models/gpt2/modeling_flax_gpt2.py:534]:
```suggestion
        if output_hidden_states:
            all_hidden_states = outputs[1] + (hidden_states,)
            outputs = (hidden_states, all_hidden_states) + outputs[2:]
        else:
            outputs = (hidden_states,) + outputs[1:]
```

The model output classes support indexing, so we don't need to check `return_dict` here

File: src/transformers/models/gpt2/modeling_flax_gpt2.py [hunk 2]
```diff
@@ -527,12 +524,23 @@ def __call__(
...
         if not return_dict:
-            return (hidden_states,) + outputs[1:]
+            if all_hidden_states:
+                return (hidden_states, all_hidden_states) + outputs[2:]
+            else:
+                return (hidden_states,) + outputs[1:]
```
Comment (patil-suraj) [inline src/transformers/models/gpt2/modeling_flax_gpt2.py:539]:
```suggestion
            return  outputs
```

File: src/transformers/models/gpt_neo/modeling_flax_gpt_neo.py [hunk 3]
```diff
@@ -557,12 +554,23 @@ def __call__(
...
         hidden_states = self.ln_f(hidden_states)
 
+        all_hidden_states = None
+        if output_hidden_states:
+            if not return_dict:
+                all_hidden_states = outputs[1]
+            else:
+                all_hidden_states = outputs.hidden_states
+            all_hidden_states = all_hidden_states + (hidden_states,)
+
```
Comment (patil-suraj) [inline src/transformers/models/gpt_neo/modeling_flax_gpt_neo.py:564]: same comment as above.

**Review by patil-suraj — APPROVED** [review]
Thanks a lot for fixing this!

**Comment by ydshieh** [comment]
@patil-suraj Thank you for the suggestions, I learned something more.

This is the first time I see "Commit suggestion" functionality. Is it OK for me to click these 3 "commit suggestion" separately, or is anything like "Commit all suggestions"?

**Comment by ydshieh** [comment]
Hi, @patil-suraj , thank you for the suggestions. There is however some issues regarding

```
        if output_hidden_states:
            all_hidden_states = outputs[1] + (hidden_states,)
            outputs = (hidden_states, all_hidden_states) + outputs[2:]
        else:
            outputs = (hidden_states,) + outputs[1:]
```
because this will change `outputs` to be a `tuple` even if it is previously a `FlaxBaseModelOutput`, and this causes problem to the end (i.e. if `return_dict=True`)

```
        return FlaxBaseModelOutput(
            last_hidden_state=hidden_states,
            hidden_states=outputs.all_hidden_states,
            attentions=outputs.attentions,
        )
```

Do you have a good solution to address this?

**Comment by patil-suraj** [comment]
Ahh, yeah, you're right!

I wanted to avoid multiple if/else conds, but seems we will need to add one either way.

I could see two options:

- we have already stored `all_hidden_states`, we could store the `all_attentions` using 

```python3
all_attentions = outputs[-1] if output_attentions else None
```

and then use that in the output class

- another option is, `FlaxGPT2BlockCollection` is only used internally, so we could also just return `outputs` (including `None` values) as a `tuple`. So in the `FlaxGPT2Module`, we could do 

```python3
if output_hidden_states:
	all_hidden_states = outputs[1] + (hidden_states,)
	outputs = (hidden_states, all_hidden_states) + outputs[2:]
else:
	outputs = (hidden_states,) + outputs[1:]

if not return_dict:
    return tuple(v for v in outputs if v is not None)

return FlaxBaseModelOutput(
    last_hidden_state=hidden_states,
    hidden_states=all_hidden_states,
    attentions=outputs[-1],
)
```

**Comment by ydshieh** [comment]
@patil-suraj  I went for option 2, with a slightly change `hidden_states=all_hidden_states,` -> `hidden_states=outputs[1],` (all_hidden_states not always define).
